### DISCUSSION TITLE
New modding advice

### DISCUSSION THREAD

**Discussion by BitForger** [slyh/Cities2-TrafficLightsEnhancement#101] [General]
Hi,

New to modding CS:2 but I used to mod Minecraft and am a current SWE. How do you recommend I go about modding CS:2? I've set up my environment following the PDX provided toolchain for context. My goal is to modify vehicle behaviors, but I'm not sure what the correct way to extend and override the functions would be.

Thanks for taking the time to look at this.

Regards,

**Comment by slyh** [comment]
I have no experience in modding Minecraft, and my last modding experience was with AMX Mod X. Therefore, I can't say how similar or dissimilar it is to Minecraft. However, you might be disappointed if you're expecting a public facing API and documentation.

Since CS2 heavily uses ECS and Burst, we can't extend or override functions as expected. But given that virtually everything in the game is based on entities and components, you can change the behaviour of anything by modifying the correct component **at the right time**. Unfortunately, there is no official documentation on components, so we have to figure out the meaning of each component ourselves.

At the beginning, I would recommend using dnSpy to decompile and modify the game's managed library. Most of the game's ECS system is inside Game.dll, and they are burst compiled to native code, so you have to disable Burst for the modified DLL to take effect.

Personally, I use a few tricks to make my life easier:

1. If you are using the Steam version, create a file called `steam_appid.txt` in the root folder of the game. The file should contain `949230`. By doing this, you can launch Cities2.exe directly without launching the Paradox Launcher.

2. You can use the command line option `--burst-disable-compilation` to disable Burst.

3. If you have [BepInEx 6 Bleeding Edge version](https://builds.bepinex.dev/projects/bepinex_be) installed, you can enable the debugging server by setting `debug_enabled` to `true` in `doorstop_config.ini`. You can use the debugger that comes with dnSpy, although it frequently crashes the game.

**Reply by BitForger** [reply]
Gotcha. I'll try that and see how far I can get. Appreciate the quick reply!
